README.md
# Login Exporter

Login Exporter is a simple Prometheus exporter that uses Chrome and 
Chrome-driver to open a website, to log in and then check for a given
text in the result. It simulate the first user interactions for the
given application. This can be used for the general availability test
of the application for the end users and for status management.

## Installation

This application it self does not need any installation, copy the binary
in the given directory should be enough. For running the logins and test
it uses chrome-headless (chromium-headless) and the chrome-driver, both
of them should be installed on the given machine. To Install them you
need to follow these guides:

### Install on MacOS

To install the needed dependencies, use brew:

```bash
brew install chromedriver
brew cask install google-chrome
```

### Install on Ubuntu

To install the needed dependencies on Ubuntu:

```bash
# Install Chrome.
sudo curl -sS -o - https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add
sudo echo "deb http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google-chrome.list
sudo apt-get -y update
sudo apt-get -y install google-chrome-stable

# Install ChromeDriver.
CHROME_DRIVER_VERSION=`curl -sS chromedriver.storage.googleapis.com/LATEST_RELEASE`
wget -N http://chromedriver.storage.googleapis.com/$CHROME_DRIVER_VERSION/chromedriver_linux64.zip -P ~/
unzip ~/chromedriver_linux64.zip -d ~/
rm ~/chromedriver_linux64.zip
sudo mv -f ~/chromedriver /usr/local/bin/chromedriver
sudo chown root:root /usr/local/bin/chromedriver
sudo chmod 0755 /usr/local/bin/chromedriver
```

## Configuration

The following parameters can be set for the application:

|  parameter  | default           | help|
|------------ |-------------------|-----|
| -config     | login.yml         | The path for the configuration file. This should be readable by running user|
| -listen_ip  | 127.0.0.1         | The IP address the application should listen to|
| -listen_port| 9980              | The port the application is listening to|
| -log_level  | INFO              | The log level for the application|
| -log_path   | login_exporter.log| The path for the log file. This should be writable by the running user|
| -timeout    | 60                | The timeout for the application to stop the check in seconds|


## Login.YML

There is an example file `login.yml.dist` available which can be used as
template for the `login.yml`. For every type of login the following 
parameters should be set:

### Simple Form

The simple is the most common type of login that is used for different
application. As long as no `CAPTCHAS` are used, this should work for every
kind of login forms. The following parameters should be set:

| parameter      | help           |
|----------------|----------------|
| login_type     | simple_form    |
| target         | The target that is searched for to find the config|
| url            | The url that login form is included|
| username       | The username that should be used for the login|
| password       | The password that should be used for the login|
| username_xpath | The xpath address of the username field (must be unique)|
| password_xpath | The xpath address of the password field (must be unique)|
| submit_xpath   | The xpath address of the submit button (must be unique)|
| expected_text  | The text that is expected to be there|

The following parameters are optional:

| parameter          | help           |
|--------------------|----------------|
| expected_text_xpath| The expected text xpath (must be unique), if not given the whole text is searched for the string|
| submit_type        | The type of submission that should be used it can be click or submit, default is submit|
| logout_xpath       | The xpath that should be used for the logout button|
| wait_time          | The time that should the form submitter wait for the page to load in seconds|
| expected_text_frame| The frame that the expected text resides|
| logout_submit_type | The submit type for the logout|
| logout_frame       | The frame that should be used for the logout|

### Shibboleth 

Shibboleth is the type used for the services that use the Shibboleth IDP
for the authentication. As all of them at the end land on the same form,
this method can be used for all of them. The following parameters must
be set for the Shibboleth:

| parameter      | help           |
|----------------|----------------|
| login_type     | shibboleth     |
| target         | The target that is searched for to find the config|
| url            | The url that login form is included|
| username       | The username that should be used for the login|
| password       | The password that should be used for the login|
| expected_text  | The text that is expected to be there|

The following parameters are optional:

| parameter          | help           |
|--------------------|----------------|
| username_xpath | The xpath address of the username field (must be unique) for Shibboleth default value is the `//input[@id='username']`|
| password_xpath | The xpath address of the password field (must be unique) for Shibboleth default value is the `//input[@id='password']`|
| submit_xpath   | The xpath address of the submit button (must be unique) for Shibboleth default value is the `//button[@class='aai_login_button']`|
| expected_text_xpath| The expected text xpath (must be unique), if not given the whole text is searched for the string|
| submit_type        | The type of submission that should be used it can be click or submit, default is submit|
| expected_text_frame| The frame that the expected text resides|
| logout_submit_type | The submit type for the logout|
| logout_frame       | The frame that should be used for the logout|


### Basic Auth

The application can be used to login in Basic Auth systems. The 
following parameters should be set for the application:

| parameter      | help           |
|----------------|----------------|
| login_type     | basic_auth     |
| url            | The url that login form is included|
| target         | The target that is searched for to find the config|
| username       | The username that should be used for the login|
| password       | The password that should be used for the login|
| expected_text  | The text that is expected to be there|

The following parameters are optional:

| parameter          | help           |
|--------------------|----------------|
| expected_text_xpath| The expected text xpath (must be unique), if not given the whole text is searched for the string|
| expected_text_frame| The frame that the expected text resides|
| logout_submit_type | The submit type for the logout|
| logout_frame       | The frame that should be used for the logout|

### Password Only

Some login fields like mailman only have a single password or token
field, this kind of authentications are also supported. The following parameters
should be set:

| parameter      | help           |
|----------------|----------------|
| login_type     | password_only     |
| url            | The url that login form is included|
| target         | The target that is searched for to find the config|
| password       | The password that should be used for the login|
| password_xpath | The xpath address of the password field (must be unique)|
| expected_text  | The text that is expected to be there|

The following parameters are optional

| parameter          | help           |
|--------------------|----------------|
| expected_text_xpath| The expected text xpath (must be unique), if not given the whole text is searched for the string|
| submit_type        | The type of submission that should be used it can be click or submit, default is submit|
| logout_xpath       | The xpath that should be used for the logout button|
| expected_text_frame| The frame that the expected text resides|
| logout_submit_type | The submit type for the logout|
| logout_frame       | The frame that should be used for the logout|

### API

This login exporter has limited support of API logins and approaches. 
The API part of the application does not use the chrome driver and uses
the `http.client` of go directly for the procedures. The following 
parameters should be set:
 
| parameter      | help           |
|----------------|----------------|
| login_type     | api     |
| url            | The url that login form is included|
| target         | The target that is searched for to find the config|
| password       | This could be key, password or token that should be used|
| password_xpath | This is the parameter that should be used for the token or password|
| expected_text  | The text that is expected to be there|

The following parameters are optional:

| parameter      | help           |
|----------------|----------------|
| username       | The username that should be used for the login|
| username_xpath | The parameter that is used for the username|
| method         | The method that should be used for the call default is POST|

### No Auth

When the page is only protected by IP address firewall or does not
have any authentication, but still should be checked for a given text
this method can be used.

| parameter      | help           |
|----------------|----------------|
| login_type     | no_auth     |
| url            | The url that login form is included|
| target         | The target that is searched for to find the config|
| expected_text  | The text that is expected to be there|

These parameters are optional:

| parameter          | help           |
|--------------------|----------------|
| expected_text_xpath| The expected text xpath (must be unique), if not given the whole text is searched for the string|

## Configuring Prometheus

This exporter works the same way as
[blackbox exporter](https://github.com/prometheus/blackbox_exporter). As
it uses a full browser to run the queries, The queries should
be done in bigger intervals and timeout should be set to a higher number.

In `prometheus.yaml the following settings should be enough:

```yaml
  - job_name: 'login_exporter'
    scrape_interval: 5m
    metrics_path: /probe
    relabel_configs:
      - source_labels: [__address__]
        target_label: __param_target
      - source_labels: [__param_target]
        target_label: instance
      - target_label: __address__
        replacement: 127.0.0.1:9980
    file_sd_configs:
    - refresh_interval: 2m
      files:
         - '/etc/prometheus/login_targets.json'
```

In `login_targets.json` the following settings would be enough:

```json
[
  {
    "labels": {
      "group": "apps",
      "host": "hostname",
      "ip": "ip_address",
      "job": "login_exporter"
    },
    "targets": [
      "target_which_is_defined_in_login.yml_before"
    ]
  }
]
```

## Development

This application is open-source and can be extended. This repository
is a mirror of our home owned repository, as a result the pull request
here can not be directly merged. But pull requests are still welcome. 
I will extract the patch and add it manually to our internal repo, when
it is acceptable patch. You still can fork this repository and add your 
changes too.

### Build

To build this application you need to install the needed requirements
first with go get. Dont forget to install chrome and chrome driver
as it was discussed above.

```bash
go get
go build -o ./login_exporter
```

Or if you want to create a binary for several platforms at once, you can
use the `go_build.sh` script.

## Change Log

See the CHANGELOG file.

## License

See the LICENSE file.

CHANGELOG
# Change Log

## [0.0.10] 2019-10-24

- Added 2 seconds for the waiting when the server is used a lot

## [0.0.9] 2019-10-23

- Removed the redundant time conversion
- Added the ability to select frames in the text search and logout
- Simplified the redundant codes

## [0.0.8] 2019-10-01

- Wait for the content to be retrieved. It is only a second.

## [0.0.7] 2019-07-10

- Dont Allow the browser to open insecure pages.

## [0.0.6] 2019-07-09

- Remove the panic from the navigation part

## [0.0.5] 2019-07-08

- Remove the no-sandbox parameter to make chrome driver killed cleanly.

## [0.0.4] 2019-06-26

- Added a possibility to wait for the page to be rendered, needed for the slow pages

## [0.0.3] 2019-06-25

- Changed the name of the variables for the metrics to have login in them to avoid collisions

## [0.0.2] 2019-06-19

- Added target as own parameter different from the url
- Added submit_type parameter which can be used for buttons or forms that can be clicked or submitted
- Some fixes in the application it self
- Allow logout to also be initiated, when the xpath exists
- Allow No Auth, when the page does not have any auth

## [0.0.1] 2019-06-18

- First release

main.go
package main

import (
	"fmt"
	log "github.com/Sirupsen/logrus"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"net/http"
)

func probeHandler(w http.ResponseWriter, r *http.Request, configs LoginConfigs) {

	// Logs all the connections
	logger.WithFields(
		log.Fields{
			"subsytem": "probe_handler",
			"part":     "connection_info",
			"user_address":       r.RemoteAddr,
			"server_host":     r.Host,
			"user_agent":    r.UserAgent(),
		}).Info("This connection was established")

	var loginType = ""
	// Extract the target from the url
	target := r.URL.Query().Get("target")
	if target == "" {
		logger.WithFields(
			log.Fields{
				"subsystem": "probe_handler",
				"part":      "target_check",
			}).Error("The target is not given")
	}
	// Find the target in the configuration
	targetConfig, err := findTargetInConfig(configs, target)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "probe_handler",
				"part":      "target_config_check",
			}).Error("The given target does not have configuration")
	} else {
		loginType = targetConfig.LoginType
	}
	// Get the status and elapsed time for the tests
	status, elapsed := getStatus(targetConfig)
	statusValue := 0
	if status {
		statusValue = 1
	}
	var statusMetric = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "login_status",
			Help: "Shows the status of the given target 0 for failure 1 for success"},
		[]string{"target", "login_type"})
	var elapsedMetric = prometheus.NewGauge(
		prometheus.GaugeOpts{Name: "login_elapsed_seconds", Help: "Shows how long it took the get the data in seconds"})
	registry := prometheus.NewRegistry()
	registry.MustRegister(statusMetric)
	registry.MustRegister(elapsedMetric)
	statusMetric.WithLabelValues(target, loginType).Set(float64(statusValue))
	elapsedMetric.Set(elapsed)
	h := promhttp.HandlerFor(registry, promhttp.HandlerOpts{})
	h.ServeHTTP(w, r)
}

/// findTargetInConfig Finds the given target in login configs
func findTargetInConfig(configs LoginConfigs, target string) (SingleLoginConfig, error) {
	for _, config := range configs.Configs {
		if config.Target == target {
			return config, nil
		}
	}
	config := SingleLoginConfig{}
	return config, fmt.Errorf("can not find target: %s", target)
}

func main() {
	loginConfig := readConfig(configFilePath)

	http.HandleFunc("/probe", func(w http.ResponseWriter, r *http.Request) {
		probeHandler(w, r, loginConfig)
	})

	logger.WithFields(
		log.Fields{
			"subsystem": "main",
			"part":      "port_setting",
		}).Info("Started Listening on " + fmt.Sprintf("%s", listenIp) + ":" + fmt.Sprintf("%v", listenPort))

	logger.Fatal(http.ListenAndServe(fmt.Sprintf("%s", listenIp)+":"+fmt.Sprintf("%v", listenPort), nil))
}


go_build.sh
#!/bin/bash
#
# GoLang cross-compile snippet for Go 1.6+ based loosely on Dave Chaney's cross-compile script:
# http://dave.cheney.net/2012/09/08/an-introduction-to-cross-compilation-with-go
#
# To use:
#
#   $ cd ~/path-to/my-awesome-project
#   $ go-build-all
#
# Features:
#
#   * Cross-compiles to multiple machine types and architectures.
#   * Uses the current directory name as the output name...
#     * ...unless you supply an source file: $ go-build-all main.go
#   * Windows binaries are named .exe.
#   * ARM v5, v6, v7 and v8 (arm64) support
#
# ARM Support:
#
# You must read https://github.com/golang/go/wiki/GoArm for the specifics of running
# Linux/BSD-style kernels and what kernel modules are needed for the target platform.
# While not needed for cross-compilation of this script, you're users will need to ensure
# the correct modules are included.
#
# Requirements:
#
#   * GoLang 1.6+ (for mips and ppc), 1.5 for non-mips/ppc.
#   * CD to directory of the binary you are compiling. $PWD is used here.
#
# For 1.4 and earlier, see http://dave.cheney.net/2012/09/08/an-introduction-to-cross-compilation-with-go
#

# This PLATFORMS list is refreshed after every major Go release.
# Though more platforms may be supported (freebsd/386), they have been removed
# from the standard ports/downloads and therefore removed from this list.
#
PLATFORMS="darwin/amd64" # amd64 only as of go1.5
PLATFORMS="$PLATFORMS windows/amd64" # arm compilation not available for Windows
PLATFORMS="$PLATFORMS linux/amd64 linux/386"

# ARMBUILDS lists the platforms that are currently supported.  From this list
# we generate the following architectures:
#
#   ARM64 (aka ARMv8) <- only supported on linux and darwin builds (go1.6)
#   ARMv7
#   ARMv6
#   ARMv5
#
# Some words of caution from the master:
#
#   @dfc: you'll have to use gomobile to build for darwin/arm64 [and others]
#   @dfc: that target expects that you're bulding for a mobile phone
#   @dfc: iphone 5 and below, ARMv7, iphone 3 and below ARMv6, iphone 5s and above arm64
#
PLATFORMS_ARM="linux"

##############################################################
# Shouldn't really need to modify anything below this line.  #
##############################################################

type setopt >/dev/null 2>&1

SCRIPT_NAME=`basename "$0"`
FAILURES=""
SOURCE_FILE=`echo $@ | sed 's/\.go//'`
CURRENT_DIRECTORY=${PWD##*/}
OUTPUT=${SOURCE_FILE:-$CURRENT_DIRECTORY} # if no src file given, use current dir name

for PLATFORM in $PLATFORMS; do
  GOOS=${PLATFORM%/*}
  GOARCH=${PLATFORM#*/}
  BIN_FILENAME="${OUTPUT}_${GOOS}_${GOARCH}"
  if [[ "${GOOS}" == "windows" ]]; then BIN_FILENAME="${BIN_FILENAME}.exe"; fi
  if [[ "${GOOS}" == "linux" ]]; then BIN_FILENAME="${BIN_FILENAME}.run"; fi
  if [[ "${GOOS}" == "darwin" ]]; then BIN_FILENAME="${BIN_FILENAME}.run"; fi
  CMD="GOOS=${GOOS} GOARCH=${GOARCH} go build -o ${BIN_FILENAME} $@"
  echo "${CMD}"
  eval $CMD || FAILURES="${FAILURES} ${PLATFORM}"
done

# ARM builds
#if [[ $PLATFORMS_ARM == *"linux"* ]]; then
#  CMD="GOOS=linux GOARCH=arm64 go build -o ${OUTPUT}-linux-arm64 $@"
#  echo "${CMD}"
#  eval $CMD || FAILURES="${FAILURES} ${PLATFORM}"
#fi
for GOOS in $PLATFORMS_ARM; do
  GOARCH="arm"
  # build for each ARM version
    BIN_FILENAME="${OUTPUT}_${GOOS}_${GOARCH}.run"
    CMD="GOOS=${GOOS} GOARCH=${GOARCH} go build -o ${BIN_FILENAME} $@"
    echo "${CMD}"
    eval "${CMD}" || FAILURES="${FAILURES} ${GOOS}/${GOARCH}"
done

# eval errors
if [[ "${FAILURES}" != "" ]]; then
  echo ""
  echo "${SCRIPT_NAME} failed on: ${FAILURES}"
  exit 1
fi

sha256sum *.run *.exe > hashes_sha256.txt

collector.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	log "github.com/Sirupsen/logrus"

	"github.com/sclevine/agouti"

	"gopkg.in/yaml.v2"
)

/// LoginConfigs Is the list of configuration
/// That should be read
type LoginConfigs struct {
	Configs []SingleLoginConfig `yaml:"targets"`
}

/// SingleLoginConfig Is the login configuration settings
/// that is used to read the yaml files.
type SingleLoginConfig struct {
	Url               string `yaml:"url"`
	Target            string `yaml:"target"`
	Username          string `yaml:"username"`
	Password          string `yaml:"password"`
	Certificate       string `yaml:"certificate"`
	UsernameXpath     string `yaml:"username_xpath"`
	PasswordXpath     string `yaml:"password_xpath"`
	CertificateXpath  string `yaml:"certificate_xpath"`
	SubmitXpath       string `yaml:"submit_xpath"`
	LoginType         string `yaml:"login_type"`
	ExpectedText      string `yaml:"expected_text"`
	ExpectedTextXpath string `yaml:"expected_text_xpath"`
	ExpectedTextFrame string `yaml:"expected_text_frame"`
	SSLCheck          bool   `yaml:"ssl_check"`
	Debug             bool   `yaml:"debug"`
	Method            string `yaml:"method"`
	SubmitType        string `yaml:"submit_type"`
	LogoutXpath       string `yaml:"logout_xpath"`
	LogoutSubmitType  string `yaml:"logout_submit_type"`
	LogoutFrame       string `yaml:"logout_frame"`
	WaitTime          int    `yaml:"wait_time"`
}

/// getChromeOptions Returns the options for the chrome driver that is used
/// to fetch the data from the server
func getChromeOptions() agouti.Option {
	return agouti.ChromeOptions("args", []string{
		"--headless",
		"--disable-gpu",
		"--no-first-run",
		"--no-default-browser-check",
		"--allow-insecure-localhost",
	})
}

/// getChromeExcludedOptions Returns the excluded options
func getChromeExcludedOptions() agouti.Option {
	return agouti.ChromeOptions("excludeSwitches", []string{
		"ignore-certificate-errors",
	})
}

///startDriver Starts the given driver on the machine
func startDriver() *agouti.WebDriver {
	driver := agouti.ChromeDriver(getChromeOptions(), getChromeExcludedOptions(), agouti.RejectInvalidSSL)
	if err := driver.Start(); err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "driver",
				"part":      "creation",
			}).Panicln(err.Error())
	}
	return driver
}

/// stopDriver Stops the given driver running. This
func stopDriver(driver *agouti.WebDriver) {
	err := driver.Stop()
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "driver",
				"part":      "destruction",
			}).Panicln(err.Error())
	}
}

/// readConfig Reads the yaml configuration from the given server
func readConfig(path string) LoginConfigs {
	var loginConfigs LoginConfigs
	yamlFile, err := ioutil.ReadFile(path)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "config_loader",
				"part":      "read_file",
			}).Panicln(err.Error())
	}
	err = yaml.Unmarshal(yamlFile, &loginConfigs)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "config_loader",
				"part":      "parse_file",
			}).Panicln(err.Error())
	}
	return loginConfigs
}

// Configuration Params
var configFilePath string
var listenIp string
var listenPort int

// Logging Params
var logPath string
var logLevel string

// Timeout settings
var timeout int

var logger = log.New()

/// getCommandLineOptions Returns the command options from the terminal
func getCommandLineOptions() {
	flag.StringVar(&configFilePath, "config", "/etc/prometheus/login.yml", "Configuration file path")
	flag.StringVar(&listenIp, "listen_ip", "127.0.0.1", "Listen IP Address")
	flag.IntVar(&listenPort, "listen_port", 9980, "Listen Port")

	flag.StringVar(&logPath, "log_file", "login_exporter.log", "Log file path")
	flag.StringVar(&logLevel, "log_level", "INFO", "Log level")

	flag.IntVar(&timeout, "timeout", 60, "Timeout in seconds")

	flag.Parse()
}

/// getLogger Returns the logger that is used to log the data
func getLogger() *log.Logger {
	f, err := os.OpenFile(logPath, os.O_APPEND|os.O_CREATE|os.O_RDWR, 0660)
	if err != nil {
		panic(err)
	}
	logger.SetFormatter(&log.JSONFormatter{})
	logger.SetOutput(f)
	parsedLevel, err := log.ParseLevel(logLevel)
	if err != nil {
		panic(err)
	}
	logger.SetLevel(parsedLevel)
	return logger
}

/// loginSimpleForm Logs in the simple for using the username, password and the submit button
func loginSimpleForm(page *agouti.Page, urlText string, usernameXpath string, passwordXpath string, submitXpath string,
	username string, password string, submitType string, wait int) {
	err := page.Navigate(urlText)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "driver",
				"part":      "navigation_error",
			}).Warningln(err.Error())
	}
	err = page.SetImplicitWait(wait * 1000)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_simple_form",
				"part":      "waiting",
			}).Warningln(err.Error())
	}
	usernameField := page.FindByXPath(usernameXpath)
	err = usernameField.SendKeys(username)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_simple_form",
				"part":      "username_field",
			}).Warningln(err.Error())
	}
	passwordField := page.FindByXPath(passwordXpath)
	err = passwordField.SendKeys(password)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_simple_form",
				"part":      "password_field",
			}).Warningln(err.Error())
	}
	submitField := page.FindByXPath(submitXpath)
	if submitType == "click" {
		err = submitField.Click()
	} else {
		err = submitField.Submit()
	}
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_simple_form",
				"part":      "submit_field",
			}).Warningln(err.Error())
	}
}

/// loginShibboleth Logs in the shibboleth system using the given username and password
func loginShibboleth(page *agouti.Page, urlText string, username string, password string, usernameXpath string,
	passwordXpath string, submitXpath string, submitType string) {
	err := page.Navigate(urlText)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "driver",
				"part":      "navigation_error",
			}).Warningln(err.Error())
	}
	if usernameXpath == "" && passwordXpath == "" && submitXpath == "" {
		usernameXpath = "//input[@id='username']"
		passwordXpath = "//input[@id='password']"
		submitXpath = "//button[@class='aai_login_button']"
	}
	usernameField := page.FindByXPath(usernameXpath)
	err = usernameField.SendKeys(username)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_shibboleth",
				"part":      "username_field",
			}).Warningln(err.Error())
	}
	passwordField := page.FindByXPath(passwordXpath)
	err = passwordField.SendKeys(password)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_shibboleth",
				"part":      "password_field",
			}).Warningln(err.Error())
	}
	if submitType == "" {
		submitType = "click"
	}
	submitField := page.FindByXPath(submitXpath)
	if submitType == "click" {
		err = submitField.Click()
	} else {
		err = submitField.Submit()
	}
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_shibboleth",
				"part":      "submit_field",
			}).Warningln(err.Error())
	}
}

/// loginApi Logs in the given API system with the given parameter for username/password and the method for the call
func loginApi(urlText string, usernameXpath string, passwordXpath string, username string,
	password string, method string) *http.Response {
	data := map[string]string{}
	if usernameXpath != "" {
		data[usernameXpath] = username
	}
	if passwordXpath != "" {
		data[passwordXpath] = password
	}
	jsonValue, err := json.Marshal(data)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_api",
				"part":      "json_marshal",
			}).Warningln(err.Error())
	}
	req, err := http.NewRequest(method, urlText, bytes.NewBuffer(jsonValue))
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_api",
				"part":      "http_request_new",
			}).Warningln(err.Error())
	} else {
		req.Header.Set("Content-Type", "application/json")
		client := &http.Client{
			Timeout: time.Duration(timeout) * time.Second,
		}
		resp, err := client.Do(req)
		if err != nil {
			logger.WithFields(
				log.Fields{
					"subsystem": "login_api",
					"part":      "http_request_do",
				}).Warningln(err.Error())
		} else {
			defer resp.Body.Close()
		}
		return resp
	}
	return nil
}

/// loginBasicAuth Logs in the given HTTPAuth page with the username and password given
func loginBasicAuth(page *agouti.Page, urlText string, username string, password string) {
	u, err := url.Parse(urlText)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_basic_auth",
				"part":      "url_parser",
			}).Warningln(err.Error())
	} else {
		urlText = u.Scheme + "://" + username + ":" + password + "@" + u.Hostname() + u.Path + "?" + u.RawQuery
	}
	err = page.Navigate(urlText)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "driver",
				"part":      "navigation_error",
			}).Warningln(err.Error())
	}
}

// logOut Logs out of the given page using the xpath that is given for the logout.
func logOut(page *agouti.Page, logoutXpath string, submitType string, framePath string) {
	if framePath != "" {
		selectFrame(framePath, page)
	}
	logoutField := page.FindByXPath(logoutXpath)
	if submitType == "click" {
		err := logoutField.Click()
		if err != nil {
			logger.WithFields(
				log.Fields{
					"subsystem": "logout",
					"part":      "click",
				}).Warningln(err.Error())
		}
	} else {
		err := logoutField.Submit()
		if err != nil {
			logger.WithFields(
				log.Fields{
					"subsystem": "logout",
					"part":      "submit",
				}).Warningln(err.Error())
		}
	}
	if framePath != "" {
		switchToRootFrame(page)
	}
}

/// loginPasswordOnly Logs in the given website with only password and the xpath to find the field
func loginPasswordOnly(page *agouti.Page, urlText string, passwordXPath string, submitXpath string, password string,
	submitType string) {
	err := page.Navigate(urlText)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "driver",
				"part":      "navigation_error",
			}).Warningln(err.Error())
	}
	passwordField := page.FindByXPath(passwordXPath)
	err = passwordField.SendKeys(password)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_password_only",
				"part":      "password_field",
			}).Warningln(err.Error())
	}
	submitField := page.FindByXPath(submitXpath)
	if submitType == "click" {
		err = submitField.Click()
	} else {
		err = submitField.Submit()
	}
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "login_password_only",
				"part":      "submit_field",
			}).Warningln(err.Error())
	}
}

/// checkExpected Checks if the expected text exists in the given path or page
func checkExpected(page *agouti.Page, expectedXPath string, expectedText string) bool {
	if expectedXPath != "" {
		expectedElement := page.FindByXPath(expectedXPath)
		expectedFieldText, err := expectedElement.Text()
		if err != nil {
			logger.WithFields(
				log.Fields{
					"subsystem": "check_expected",
					"part":      "xpath_data",
				}).Warningln(err.Error())
			return false
		}
		return expectedFieldText == expectedText
	}
	time.Sleep(2 * time.Second)
	content, err := page.HTML()
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "check_expected",
				"part":      "match_all_data",
			}).Warningln(err.Error())
		return false
	}
	return strings.Contains(content, expectedText)
}

/// getNoLogin The no login function that only returns the page without any submissions
func getNoLogin(page *agouti.Page, urlText string) {
	err := page.Navigate(urlText)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "driver",
				"part":      "navigation_error",
			}).Warningln(err.Error())
	}
}

/// checkExpectedResponse Checks if the expected string exists in the http.Response
func checkExpectedResponse(response *http.Response, expectedText string) bool {
	content, err := ioutil.ReadAll(response.Body)
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "check_expected",
				"part":      "get_content",
			}).Warningln(err.Error())
	}
	return strings.Contains(string(content), expectedText)
}

/// selectFrame selects given frame for the given page
func selectFrame(framePath string, page *agouti.Page) {
	err := page.Find(framePath).SwitchToFrame()
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "check_expected",
				"part":      "switch",
			}).Warningln(err.Error())
	}
}

/// switchToRootFrame Switches to the root frame
func switchToRootFrame(page *agouti.Page) {
	err := page.SwitchToRootFrame()
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "check_expected",
				"part":      "switch_to_root_frame",
			}).Warningln(err.Error())
	}
}

///getStatus Returns the data from the server
func getStatus(config SingleLoginConfig) (status bool, elapsed float64) {
	status = false
	driver := startDriver()
	page, err := driver.NewPage()
	if err != nil {
		logger.WithFields(
			log.Fields{
				"subsystem": "driver",
				"part":      "new_page",
			}).Warningln(err.Error())
	} else {
		err = page.SetPageLoad(timeout * 1000)
	}

	start := time.Now()
	switch config.LoginType {
	case "simple_form":
		loginSimpleForm(page, config.Url, config.UsernameXpath, config.PasswordXpath, config.SubmitXpath,
			config.Username, config.Password, config.SubmitType, config.WaitTime)
		break
	case "shibboleth":
		loginShibboleth(page, config.Url, config.Username, config.Password, config.UsernameXpath,
			config.PasswordXpath, config.SubmitXpath, config.SubmitType)
		break
	case "basic_auth":
		loginBasicAuth(page, config.Url, config.Username, config.Password)
		break
	case "password_only":
		loginPasswordOnly(page, config.Url, config.PasswordXpath, config.SubmitXpath, config.Password, config.SubmitType)
		break
	case "no_auth":
		getNoLogin(page, config.Url)
		break
	}
	if config.LoginType == "api" {
		response := loginApi(config.Url, config.UsernameXpath, config.PasswordXpath, config.Username, config.Password,
			config.Method)
		status = checkExpectedResponse(response, config.ExpectedText)
	} else {
		if config.ExpectedTextFrame != "" {
			selectFrame(config.ExpectedTextFrame, page)
		}
		status = checkExpected(page, config.ExpectedTextXpath, config.ExpectedText)
		if config.ExpectedTextFrame != "" {
			switchToRootFrame(page)
		}
	}
	end := time.Now()
	elapsed = end.Sub(start).Seconds()
	// logout if the value is set
	if config.LogoutXpath != "" {
		logOut(page, config.LogoutXpath, config.LogoutSubmitType, config.LogoutFrame)
	}

	if err == nil {
		err = page.CloseWindow()
		if err != nil {
			logger.WithFields(
				log.Fields{
					"subsystem": "driver",
					"part":      "new_page_close_error",
				}).Warningln(err.Error())
		}
	}
	stopDriver(driver)
	return status, elapsed
}

func init() {
	getCommandLineOptions()
	getLogger()
}
